compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -1626,8 +1626,26 @@ impl<'a> Parser<'a> {
     }
 
     /// Parses an optional `move` prefix to a closure-like construct.
-    fn parse_capture_clause(&mut self) -> CaptureBy {
-        if self.eat_keyword(kw::Move) { CaptureBy::Value } else { CaptureBy::Ref }
+    fn parse_capture_clause(&mut self) -> PResult<'a, CaptureBy> {
+        if self.eat_keyword(kw::Move) {
+            // Check for `move async` and recover
+            if self.check_keyword(kw::Async) {
+                let move_async_span = self.token.span.with_lo(self.prev_token.span.data().lo);
+                let mut diagnostics =
+                    self.struct_span_err(move_async_span, "unexpected order: `move async`");
```

Comment:
Could you change the diagnostic message to the following:

```suggestion
                    self.struct_span_err(move_async_span, "the order of `move` and `async` is incorrect");
```

This just keeps it consistent with other similar diagnostics:
https://github.com/rust-lang/rust/blob/0d97f7a96877a96015d70ece41ad08bb7af12377/src/test/ui/mut/mut-ref.stderr#L1